File: Algorithms/C++/00374-Guess Number Higher or Lower.cpp
/** 
 * Forward declaration of guess API.
 * @param  num   your guess
 * @return 	     -1 if num is higher than the picked number
 *			      1 if num is lower than the picked number
 *               otherwise return 0
 * int guess(int num);
 */

class Solution {
public:
    int guessNumber(int n) {
        int start{1};
        int end{n};
        while(start <= end){
            int mid = start + (end - start)/2;
            int myGuess = guess(mid);
            if(myGuess == 1){
                start = mid+1;
            } else if(myGuess == -1){
                end = mid-1;
            } else{
                return mid;
            }
        }

        return -1;
    }
};

/*
    mid < target : 1
    mid > target : 1
    mid == target : 0

    [1,2,3,4,5]

    mid = (start+end)/2
    if(mid > target){
        start = mid+1
    } else if(mid < target){
        end = mid-1
    }
*/

File: Algorithms/Python3/00374-Guess Number Higher or Lower.py
# The guess API is already defined for you.
# @param num, your guess
# @return -1 if num is higher than the picked number
#          1 ifnum is lower than the picked number
#          otherwise return  0
# def guess(num: int) -> int:

class Solution:
    def guessNumber(self, n: int) -> int:
        result = 2
        l = 1
        r = n
        while result != 0:
            mid = l + ((r-l)/2)
            result = guess(mid)
            if result == -1:
                r = mid-1
            elif result == 1:
                l = mid+1
        
        return int(mid)

'''
[1,2,3,4,5]

(1 + 5 / 2) = 3

-1: num > target
1: num < target
0: num == target

'''